File: sast-java/src/main/java/com/sast/astbenchmark/case_language_maturity/accuracy/context_sensitive/argument_return_value_passing/ChainedCall_002_F.java
package com.sast.astbenchmark.case_language_maturity.accuracy.context_sensitive.argument_return_value_passing;

import com.sast.astbenchmark.common.utils.SinkUtil;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.Map;

// evaluation information start
// real case = false
// evaluation item = 准确度->上下文敏感分析->参数/返回值传递
// scene introduction = 链式调用
// level = 2
// bind_url = accuracy/context_sensitive/argument_return_value_passing/ChainedCall_002_F
// date = 2025-12-10 15:15:00
// evaluation information end
@RestController()
@RequestMapping("accuracy/context_sensitive/argument_return_value_passing")
public class ChainedCall_002_F {
  @PostMapping("ChainedCall_002_F")
  public Map<String, Object> chainedcall_002_f(@RequestParam String cmd) {
    Map<String, Object> modelMap = new HashMap<>();
    try {
      // 场景特点：链式调用中的返回值传递，非污点通过链式调用传递
      String result = new CommandProcessor().setCommand("safe_value").process();
      Runtime.getRuntime().exec(result);
      modelMap.put("status", "success");
    } catch (Exception e) {
      modelMap.put("status", "error");
    }
    return modelMap;
  }

  static class CommandProcessor {
    private String command;

    public CommandProcessor setCommand(String command) {
      this.command = command;
      return this;
    }

    public String process() {
      return command;
    }
  }
}

File: sast-php/case/completeness/single_app_tracing/function_call/chained_call_002_F.php
<?php
// evaluation information start
// real case = false
// evaluation item = 完整度->单应用追踪->函数调用
// scene introduction = 链式调用最终输出安全常量
// level = 2
// bind_url = completeness/single_app_tracing/function_call/chained_call_002_F
// evaluation information end

class ConstPipeline {
    private $value = "";

    public function input($val) {
        $this->value = $val;
        return $this;
    }

    public function output() {
        return "constant_output";
    }
}

function chained_call_002_F($__taint_src) {
    $pipe = new ConstPipeline();
    $result = $pipe->input($__taint_src)->output();
    __taint_sink($result);
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
chained_call_002_F($taint_src);
